src/core/GAT.jl

```diff
@@ -589,34 +627,34 @@ the Typeclass.jl library for Julia.
 """
 macro instance(head, body)
   # Parse the instance definition.
-  head = parse_signature_binding(head)
-  functions = parse_instance_body(body)
+  head = parse_theory_binding(head)
+  functions, ext_functions = parse_instance_body(body)
 
-  # We must generate and evaluate the code at *run time* because the signature
+  # We must generate and evaluate the code at *run time* because the theory
   # module is not defined at *parse time*.
   # Also, we "throw away" any docstring.
   # FIXME: Is there a better place to put the docstring?
-  expr = :(instance_code($(esc(head.name)), $(esc(head.params)), $functions))
+  expr = :(instance_code($(esc(head.name)), $(esc(head.params)), $functions, $ext_functions))
   Expr(:block,
     Expr(:call, esc(:eval), expr),
     :(Core.@__doc__ abstract type $(esc(gensym(:instance_doc))) end)) # /dev/null
 end
-function instance_code(mod, instance_types, instance_fns)
+function instance_code(mod, instance_types, instance_fns, external_fns)
   code = Expr(:block)
   class = mod.class()
   bindings = Dict(zip(class.type_params, instance_types))
   bound_fns = [ replace_symbols(bindings, f) for f in interface(class) ]
   bound_fns = OrderedDict(parse_function_sig(f) => f for f in bound_fns)
   instance_fns = Dict(parse_function_sig(f) => f for f in instance_fns)
-  for (sig, f) in bound_fns
-    if haskey(instance_fns, sig)
-      f_impl = instance_fns[sig]
+  for (theory, f) in bound_fns
+    if haskey(instance_fns, theory)
+      f_impl = instance_fns[theory]
     elseif !isnothing(f.impl)
       f_impl = f
-    else
-      # FIXME: Can we check whether the method has already been defined?
-      @warn "Method $(f.call_expr) not implemented in $(class.name) instance"
+    elseif f.call_expr.args[1] in external_fns
       continue
+    else
+      @error "Method $(f.call_expr) not implemented in $(class.name) instance"
```

Comment:
Rather than logging an error using the `Logging` module, I think we should actually raise an exception. This was the original behavior, which I changed to a warning as a temporary workaround.

Reply:
Fixed in commit 1a3a27c